source/probability_theory_4_finite.Rmd

```diff
@@ -524,284 +481,178 @@ sampling mechanism as in Example 8-1, but now we record "yes" for a
-determine whether a match occurs. Our method is using the SUBTRACT
-command to compare each element of A with each element of C. A match
-causes a "0," this would mean that the third couple was matched
-correctly because they have the same numbers---"3"s. We then COUNT the
-number of "0"s in D to determine the number of couples correctly matched
-in the current sample. The result for the current sample is placed in E,
-then transferred to Z in order to keep SCORE for each of the 1000
-samples. This ends the loop.
-
-`           NUMBERS 1,10 a         `
-
-An array of 10 males.
-
-`           NUMBERS 1,10 b         `
-
-An identical array of 10 females---the pair for each of the males.
-
-`           REPEAT 1000         `
-
-Do the experiment 1000 times.
+numbers.  The solution follows below.
 
-`           SHUFFLE b c         `
-
-Shuffle the females.
-
-`           END         `
-
-`           SUBTRACT a c d         `
-
-This operation pairs each shuffled female with a male and subtracts. If
-it is an original pairing (1/1, 2/2, etc.), the result will be a 0. The
-number of 0's indicates how many of the 10 males got paired up again
-with their original partner.
-
-`           COUNT d =0 e         `
-
-Count the number of 0's and put the result in e.
-
-`           SCORE e z         `
-
-Keep track of each trial result.
-
-End the trial, go back and repeat until all 1000 are complete.
-
-`           HISTOGRAM z         `
-
-Produce a histogram of the results.
+```{python, opts.label="py_ed"}
+N = 10000
+trial_results = np.zeros(N, dtype=int)
 
-![](images/12-Chap-8_007.png)
+# Assign numbers to each student, according to their pair, after the first
+# icebreaker
+birmingham = np.arange(10)
+berkeley = np.arange(10)
 
-From this histogram, we see that in about 3 percent of the trials did 4
-or more couples end up being re-paired with their own partners. We can
-calculate this proportion directly with RESAMPLING STATS:
+for i in range(N):
+    # Randomly shuffle the students from berkeley
+    # (We could also have shuffled both arrays---it does not alter the outcome)
+    shuffled_berkeley = np.random.permutation(berkeley)
 
-`           COUNT z >= 4 k         `
+    # Count in how many cases people landed with the same person as in the
+    # first round, and store in trial_results.
+    matches = np.count_nonzero(shuffled_berkeley == birmingham)
+    trial_results[i] = matches
 
-Determine how many trials had 4 or more males being matched with their
-partner after the shuffling. (Note that this is the same as the number
-of females being matched with their original partner.)
+# Count the number of times we got 4 or more people assigned to the same person
+k = np.count_nonzero(trial_results >= 4)
 
-`           DIVIDE k 1000 kk         `
+# Convert to a proportion.
+kk = k / N
 
-Convert to a proportion.
+# Print the result.
+print(kk)
+```
 
-`           PRINT kk         `
+**NOTE:** This seems to be 1.5 percent, instead of 3 now?
```

Comment:
I can't see the problem in your code, I agree.  I'll try and run the program in resampling stats, I think I have the code somewhere.

Reply:
Could be that his N was too low.

Reply:
Yup, could well be.  Still, I'll see if I can resurrect the code just in case.

Reply:
I get 0.0189 with 1000000 samples in resampling stats, which is very close to what I get running the notebook here.